### PR TITLE
feat(eslint-plugin): [safe-array-destructring] Restrict destructuring

### PR DESCRIPTION
**Pull request by nicu-chiciuc**
I've copied the initial draft of the markdown explanation from `safe-array-destructuring.md` below since, in case it will be approved, I expect it to be shorter and not contain as many references. As it is right now I think it thoroughly explains the reasoning behind this rule.

### Restricts destructuring of tuple/arrays that may be empty (`safe-array-destructuring`)

Allow array destructuring to only be applied to tuple types, excluding tuple rest parameters.

The following Typescript code is valid even though it is clearly incorrect:

```ts
const arr: number[] = [];
const [first, second] = arr;
```

The issue is that array indexing is not defined in a type-safe way (as it usually happens in most known type systems).
This issue is discussed in [TS issue #13778](https://github.com/microsoft/TypeScript/issues/13778).
An approach to solve this is to change the definition of Array indexing in `lib.es5.d.ts`,
from:

```ts
interface Array<T> {
  // ...

  [n: number]: T;
}
```

to

```ts
interface Array<T> {
  // ...

  [n: number]: T | undefined;
}
```

This can be automated using something like [`patch-package`](https://github.com/ds300/patch-package) or some other postinstall script.

The problem that this rule aims to solve is that for `target` versions other than `es5` the approach with overriding the `Array` interface does not apply to array destructuring for some reason [TS issue #37045](https://github.com/microsoft/TypeScript/issues/37045).

Irregardless, this rule might be of use in projects where array destructuring needs to be more strictly enforced.

Examples of **incorrect** code for this rule:

```ts
const array: number[] = [];
const [first] = array;
```

```ts
const inner: { a: number[] } = { a: [] };
const {
  a: [value],
} = inner;
```

Examples of **correct** code for this rule:

```ts
const restTuple: [number, ...number[]] = [3];
const [first, ...rest] = restTuple;
```

```ts
type SemiTuple = [string, ...number[]];
const inner: { a: SemiTuple } = { a: ['test'] };

const {
  a: [value],
} = inner;
```

### Related To

- Typescript array [indexing issue](https://github.com/microsoft/TypeScript/issues/13778)
- TypeScript [`target` option issue](https://github.com/microsoft/TypeScript/issues/37045)

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/safe-array-destructuring.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+const inner: { a: SemiTuple } = { a: ['test'] };
+
+const {
+  a: [value],
+} = inner;
+```
+
+## Related To
+
+- Typescript array [indexing issue](https://github.com/ds300/patch-package)
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/safe-array-destructuring.md]: incorrect link?

Comment (nicu-chiciuc) [reply packages/eslint-plugin/docs/rules/safe-array-destructuring.md]: Fixed

**Comment by bradzacher** [comment]
The only problem I have with this rule is it means there's no nice way to work around the rule without disabling the rule with a comment.

```ts
function foo(input: number[]) {
  if (input.length > 3) {
    const [a,b,c] = input; // lint error because type is still number[]
  }
}
```

I guess you can work around it like this, but it's pretty cumbersome for higher lengths.
```ts
function foo(input: number[]) {
  if (input.length > 3) {
    const [a, b, c] = input as [number, number, number, ...number[]];
  }
}
```

I wonder if there's any easy way for this rule to do some basic control-flow analysis to help with this?

**Comment by nicu-chiciuc** [comment]
That would be a nice feature, although I guess the enforcement should be at the type level. I'm not sure if `typescript-eslint` can act as a sort of type-guard creator.
I would probably use some thing like:
```ts
function hasFirstElements<T>(n: 1, arr: T[]): arr is [T, ...T[]];
function hasFirstElements<T>(n: 2, arr: T[]): arr is [T, T, ...T[]];
function hasFirstElements<T>(n: 3, arr: T[]): arr is [T, T, T, ...T[]];
function hasFirstElements<T>(n: number, arr: T[]): arr is T[];
function hasFirstElements<T>(n: number, arr: T[]): arr is T[] {
    return arr.length === n;
};

const arr = [1, 2, 3, 4];
if (hasFirstElements(2, arr)) {
    // has type [number, number, ...number[]]
    arr;
}
```

Working with tuples created other similar problems, for example:
```
declare const arr2: [number, ...number[]];
const arr3 = [...arr2];
```

**Comment by nicu-chiciuc** [comment]
Tuples also break when using the default Array methods:
```ts
declare const arr2: [number, ...number[]];
// Has type string[]
const arr4 = arr2.map(val => val.toString());
```

I think if we continue to abuse tuples in our project I should create some public utility package to work with them while maintaining type guarantees.

**Comment by nicu-chiciuc** [comment]
After having to remove all the TS error after making the transition to safer type indexing I had to add some useful type-guards:
```ts
export type Array1<T> = [T, ...T[]];

export function map<T, U>(arr: [T, ...T[]], callbackfn: (value: T, index: number) => U): [U, ...U[]];
export function map<T, U>(arr: [T, T, ...T[]], callbackfn: (value: T, index: number) => U): [U, U, ...U[]];
export function map<T, U>(arr: T[], callbackfn: (value: T, index: number) => U): U[] {
  return arr.map(callbackfn);
}

export function hasAtLeast<T>(n: 1, arr: T[]): arr is [T, ...T[]];
export function hasAtLeast<T>(n: 2, arr: T[]): arr is [T, T, ...T[]];
export function hasAtLeast<T>(n: 3, arr: T[]): arr is [T, T, T, ...T[]];
export function hasAtLeast<T>(n: number, arr: T[]): arr is T[];
export function hasAtLeast<T>(n: number, arr: T[]): arr is T[] {
  return arr.length >= n;
}

export function hasExactly<T>(n: 1, arr: T[]): arr is [T];
export function hasExactly<T>(n: 2, arr: T[]): arr is [T, T];
export function hasExactly<T>(n: 3, arr: T[]): arr is [T, T, T];
export function hasExactly<T>(n: number, arr: T[]): arr is T[];
export function hasExactly<T>(n: number, arr: T[]): arr is T[] {
  return arr.length === n;
}
// ... and so on
```

I think this rule should not be enabled by default and if someone decides to go this route, some version of the above functions should be used. It would make sense to publish them as a separate package. I might look into that after figuring out all the pain points.

**Comment by nicu-chiciuc** [comment]
Will this rule be still taken into consideration or should I make other amendments?

**Comment by bradzacher** [comment]
Your PR is in the review queue.
It will get reviewed in due course.

https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#addressing-feedback-and-beyond

**Comment by bradzacher** [comment]
I've finally found some time to address the PR queue. Thinking about this rule more.

In its current iteration, I don't think that it adds enough value to be merged into this core plugin because it's too restrictive; it is impossible to use without:
- a disable comment,
  - this is ergonomically terrible for a lint rule, and is a non-starter.
- a manual and verbose assertion,
  - type assertions are not great because [they are somewhat unsafe, and overly permissive](https://www.typescriptlang.org/play/#code/MYewdgzgLgBAHjAvDA2gRgDQwExYMwC6MAhhDKJFANwBQF0MAnkvCWelrjHlgHT9gArgFsARgFMATigIFa9WAC8WzUqkw58WACxyadcA2Is4KQlRgB6SzCmSQkmIFByA5RiiVZuVZtRGAB3EQADMkRGQhMSkYQBlyVwZgFkUvC2sYP0CQsORtWJogA).
- creating utility functions
  - these are a sub-optimal solution, due to:
    - the user overhead of creating and importing the functions in their codebase.
    - the runtime cost of creating a closure for the function call.

I think that in order to consider this rule, it would need to do some analysis of the variable usage to determine if there has been a check done.

i.e. the following code should be valid

```ts
declare const arr: number[];
if (arr.length > 5) {
  const [a, b, c, d, e] = arr;
}
```

Though this will be very hard to do well. It's not impossible, but it will be hard.
For now, I'm going to close this. If you want to work on this further, feel free to do so and comment back, and we can reassess then.

Thanks for your work.

**Comment by nicu-chiciuc** [comment]
Thanks for the feedback, I realized from the get-go that this is a rule for a very specific use-case and might not have its place in the main plugin, since it relies on the fact that, in order for this rule to make any sense, the developer has to amend the indexing definition of arrays. I think a much better approach would be to create a custom plugin that would provide these types of rules.

I didn't put the rule in a custom plugin since I realized that would take more time than using the current project for testing and documentation.

**Pull request closed without merging**